File: mip_reference/dotnet-aspnet/Mip/Signature.cs
namespace Mip;

/// <summary>
/// MIP request signature creation and verification.
/// Per spec: signature signs "timestamp + path + json_payload"
/// </summary>
public static class Signature
{
    /// <summary>
    /// Create a MIP request signature.
    /// </summary>
    public static string SignRequest(string privateKeyPem, string timestamp, string path, string? jsonBody = null)
    {
        var data = BuildSignatureData(timestamp, path, jsonBody);
        return Crypto.Sign(privateKeyPem, data);
    }

    /// <summary>
    /// Verify a MIP request signature.
    /// </summary>
    public static bool VerifyRequest(string publicKeyPem, string signature, string timestamp, string path, string? jsonBody = null)
    {
        var data = BuildSignatureData(timestamp, path, jsonBody);
        return Crypto.Verify(publicKeyPem, signature, data);
    }

    /// <summary>
    /// Check if timestamp is within acceptable window (±5 minutes).
    /// </summary>
    public static bool TimestampValid(string timestamp, int windowSeconds = 300)
    {
        if (!DateTime.TryParse(timestamp, out var requestTime))
            return false;

        var now = DateTime.UtcNow;
        var diff = Math.Abs((now - requestTime.ToUniversalTime()).TotalSeconds);
        return diff <= windowSeconds;
    }

    private static string BuildSignatureData(string timestamp, string path, string? jsonBody)
    {
        var data = timestamp + path;
        if (!string.IsNullOrEmpty(jsonBody))
            data += jsonBody;
        return data;
    }
}

File: mip_reference/php-symfony/src/Mip/Signature.php
<?php

namespace App\Mip;

/**
 * MIP request signature handling
 */
class Signature
{
    /**
     * Create a MIP request signature
     * Per spec: signature signs "timestamp + path + json_payload"
     */
    public static function signRequest(string $privateKeyPem, string $timestamp, string $path, ?string $jsonBody = null): string
    {
        $data = self::buildSignatureData($timestamp, $path, $jsonBody);
        return Crypto::sign($privateKeyPem, $data);
    }

    /**
     * Verify a MIP request signature
     */
    public static function verifyRequest(string $publicKeyPem, string $signature, string $timestamp, string $path, ?string $jsonBody = null): bool
    {
        $data = self::buildSignatureData($timestamp, $path, $jsonBody);
        return Crypto::verify($publicKeyPem, $signature, $data);
    }

    /**
     * Check if timestamp is within acceptable window (default ±5 minutes)
     */
    public static function timestampValid(string $timestamp, int $windowSeconds = 300): bool
    {
        try {
            $requestTime = new \DateTime($timestamp);
            $now = new \DateTime();
            $diff = abs($now->getTimestamp() - $requestTime->getTimestamp());
            return $diff <= $windowSeconds;
        } catch (\Throwable $e) {
            return false;
        }
    }

    private static function buildSignatureData(string $timestamp, string $path, ?string $jsonBody): string
    {
        $data = $timestamp . $path;
        if ($jsonBody !== null && $jsonBody !== '') {
            $data .= $jsonBody;
        }
        return $data;
    }
}
